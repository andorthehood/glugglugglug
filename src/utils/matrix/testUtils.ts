export function roundValue(value: number, precision = 6) {
	const factor = 10 ** precision;
	const rounded = Math.round(value * factor) / factor;
	return Object.is(rounded, -0) ? 0 : rounded;
}

export function formatMatrix(values: number[]) {
	return Array.from({ length: 4 }, (_, row) => {
		const start = row * 4;
		return values.slice(start, start + 4).map((value) => roundValue(value));
	});
}

export function formatVector(values: number[]) {
	return values.map((value) => roundValue(value));
}
