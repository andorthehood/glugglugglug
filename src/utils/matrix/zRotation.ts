/**
 * Produces a matrix representing rotation around the z axis.
 * @param angleInRadians The rotation angle in radians.
 * @param dst Destination matrix to write into. When omitted a new matrix is created.
 * @returns A rotation matrix that rotates around the z axis.
 */
export function zRotation(
	angleInRadians: number,
	dst: number[] = new Array(16),
) {
	const c = Math.cos(angleInRadians);
	const s = Math.sin(angleInRadians);

	dst[0] = c;
	dst[1] = s;
	dst[2] = 0;
	dst[3] = 0;
	dst[4] = -s;
	dst[5] = c;
	dst[6] = 0;
	dst[7] = 0;
	dst[8] = 0;
	dst[9] = 0;
	dst[10] = 1;
	dst[11] = 0;
	dst[12] = 0;
	dst[13] = 0;
	dst[14] = 0;
	dst[15] = 1;
	return dst;
}
