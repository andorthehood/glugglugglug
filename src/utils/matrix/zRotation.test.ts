import { describe, expect, it } from "vitest";
import { zRotation } from "./zRotation";
import { formatMatrix } from "./testUtils";

describe("zRotation", () => {
	it("creates a rotation matrix around the z axis", () => {
		const angle = Math.PI / 4;

		expect(formatMatrix(zRotation(angle))).toMatchSnapshot();
	});

	it("writes into an existing matrix", () => {
		const angle = Math.PI / 4;
		const dst = new Array<number>(16);
		const result = zRotation(angle, dst);

		expect(result).toBe(dst);
		expect(formatMatrix(dst)).toMatchSnapshot();
	});
});
