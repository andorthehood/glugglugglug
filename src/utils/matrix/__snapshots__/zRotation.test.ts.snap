// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`zRotation > creates a rotation matrix around the z axis 1`] = `
[
  [
    0.707107,
    0.707107,
    0,
    0,
  ],
  [
    -0.707107,
    0.707107,
    0,
    0,
  ],
  [
    0,
    0,
    1,
    0,
  ],
  [
    0,
    0,
    0,
    1,
  ],
]
`;

exports[`zRotation > writes into an existing matrix 1`] = `
[
  [
    0.707107,
    0.707107,
    0,
    0,
  ],
  [
    -0.707107,
    0.707107,
    0,
    0,
  ],
  [
    0,
    0,
    1,
    0,
  ],
  [
    0,
    0,
    0,
    1,
  ],
]
`;
